### PR TITLE
OPENMRS-75: Support for *AttributeType.

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket : https://mekomsolutions.atlassian.net/browse/OPENMRS-75

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	protected static String HEADER_DATATYPE_CONFIG = "datatype config";
+	
+	protected static String HEADER_PREFERRED_HANDLER_CLASSNAME = "preferred handler classname";
+	
+	protected static String HEADER_HANDLER_CONFIG = "handler config";
+	
+	protected static String HEADER_DOMAIN = "domain";
+	
+	// Supported Domain Names
+	private static final String LOCATION_DOMAIN = "Location";
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java]: Refer to the already existing enum to obtain that domain name.

File: api/src/main/java/org/openmrs/module/initializer/Domain.java [hunk 2]
```diff
@@ -16,7 +16,8 @@
 	PERSON_ATTRIBUTE_TYPES(12, "personattributetypes"),
 	IDENTIFIER_SOURCES(13, "idgen"),
 	DRUGS(14, "drugs"),
-	ORDER_FREQUENCIES(15, "orderfrequencies");
+	ORDER_FREQUENCIES(15, "orderfrequencies"),
+	ATTRIBUTE_TYPE(16, "attributetypes");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/Domain.java]: `ATTRIBUTE_TYPES`

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeType.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.openmrs.module.initializer.api.attribute.type;
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeType.java]: Here again it would be rather plural (look at the program-related package names for example):
```java
package org.openmrs.module.initializer.api.attributes.types;
```

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeTypeServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		}
+		if (instance instanceof VisitAttributeType) {
+			VisitAttributeType visitAttributeType = (VisitAttributeType) instance;
+			return Context.getVisitService().saveVisitAttributeType(visitAttributeType);
+		}
+		if (instance instanceof ProviderAttributeType) {
+			ProviderAttributeType providerAttributeType = (ProviderAttributeType) instance;
+			return Context.getProviderService().saveProviderAttributeType(providerAttributeType);
+		}
+		return null;
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeTypeServiceImpl.java]: Before returning null we will need a very exhaustive and explicit error message to be logged. Something like
>this attribute type \<name here\> is not supported, here are the currently supported attribute types \<loop through the enum\>.

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeTypeServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+				return Context.getLocationService().getLocationAttributeTypeByUuid(uuid);
+			
+			case VISIT:
+				return Context.getVisitService().getVisitAttributeTypeByUuid(uuid);
+			
+			case PROVIDER:
+				return Context.getProviderService().getProviderAttributeTypeByUuid(uuid);
+			
+			default:
+				return null;
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/AttributeTypeServiceImpl.java]: Same remark here.

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	private static final String VISIT_DOMAIN = "Visit";
+	
+	private static final String PROVIDER_DOMAIN = "Provider";
+	
+	public BaseAttributeTypeLineProcessor(String[] headerLine, AttributeTypeService service) {
+		super(headerLine, service);
+	}
+	
+	public static AttributeType getAttributeType(CsvLine line) {
+		String attributeDomain = line.getString(HEADER_DOMAIN);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java]:
```java
line.getString(HEADER_DOMAIN, true)
```

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			
+		}
+		
+	}
+	
+	@SuppressWarnings("rawtypes")
+	@Override
+	protected BaseAttributeType bootstrap(CsvLine line) throws IllegalArgumentException {
+		String uuid = getUuid(line.asLine());
+		BaseAttributeType attributeType = Context.getService(AttributeTypeService.class).getAttributeType(uuid,
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java]: Look at other line processors, the ad-hoc service should be a construction parameter, it should not be statically gotten from `Context`.

File: api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		    getAttributeType(line));
+		if (attributeType == null) {
+			attributeType = newAttributeType(line);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/attribute/type/BaseAttributeTypeLineProcessor.java]: Surely you should be able to bootstrap by name if no UUID was provided?

File: api/src/main/java/org/openmrs/module/initializer/api/loaders/AttributeTypeLoader.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import java.io.InputStream;
+
+import org.openmrs.module.initializer.Domain;
+import org.openmrs.module.initializer.api.CsvParser;
+import org.openmrs.module.initializer.api.attribute.type.AttributeTypeCsvParser;
+import org.openmrs.module.initializer.api.attribute.type.AttributeTypeService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/loaders/AttributeTypeLoader.java]: As discussed this will most likely be conditional to Core 2.2.x.

File: api/src/main/resources/moduleApplicationContext.xml [hunk 10]
```diff
@@ -43,6 +43,23 @@
             <ref bean="transactionAttributeSource" />
         </property>
     </bean>
+    
+    <bean id="initializer.AttributeTypeService"
```
Comment (mks-d) [inline api/src/main/resources/moduleApplicationContext.xml]: This doesn't need to be a fully-fledged service like that, just make it a bean with `@Component` in the class definition.

Comment (samuelmale) [reply api/src/main/resources/moduleApplicationContext.xml]: Meaning we also don't need the interface `AttributeTypeService`. That's my guess!

Comment (mks-d) [reply api/src/main/resources/moduleApplicationContext.xml]: Correct, we just need the implementation class for the sake of Iniz.

File: api/src/main/java/org/openmrs/module/initializer/api/loaders/AttributeTypeLoader.java (hunk 9 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/loaders/AttributeTypeLoader.java]: > conditional to Core 2.2.x.
Surely I don't really understand that. Are you able enlighten me a little @mks-d ?

Comment (mks-d) [reply api/src/main/java/org/openmrs/module/initializer/api/loaders/AttributeTypeLoader.java]: Perhaps should you have a look at this page: https://wiki.openmrs.org/display/docs/Supporting+different+OpenMRS+versions

File: api-2.2/.classpath [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (samuelmale) [inline api-2.2/.classpath]: Will definitely remove this!

**Comment by samuelmale** [comment]
I'm wondering as to why Travis is failing to build with a compilation error!

@mks-d any idea?

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java [hunk 12]
```diff
@@ -9,11 +9,11 @@
...
 import org.openmrs.module.initializer.api.CsvLine;
 import org.openmrs.module.initializer.api.attributes.types.AttributeType;
-import org.openmrs.module.initializer.api.attributes.types.AttributeTypeLineProcessorCompatibilty;
+import org.openmrs.module.initializer.api.attributes.types.AttributeTypeCsvLineHandler;
 import org.openmrs.module.initializer.api.attributes.types.BaseAttributeTypeLineProcessor;
 
-@OpenmrsProfile(openmrsPlatformVersion = "2.2.0")
-public class AttributeTypeLineProcessorCompatibilty2_2 implements AttributeTypeLineProcessorCompatibilty {
+@OpenmrsProfile(openmrsPlatformVersion = "[2.2.0 - 2.3.*]")
+public class AttributeTypeCsvLineHandlerImpl2_2 implements AttributeTypeCsvLineHandler {
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java]: This should be
```java
AttributeTypeCsvLineHandlerImpl2_2 extends AttributeTypeCsvLineHandlerImpl1_1 {
```

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java [hunk 13]
```diff
@@ -15,15 +15,23 @@
...
 import org.openmrs.attribute.BaseAttributeType;
 import org.openmrs.module.initializer.api.attributes.types.AttributeType;
-import org.openmrs.module.initializer.api.attributes.types.AttributeTypeServiceCompatibility;
+import org.openmrs.module.initializer.api.attributes.types.AttributeTypeService;
 
-@OpenmrsProfile(openmrsPlatformVersion = "2.2.0")
-public class AttributeTypeServiceCompatibilityHandler2_2 implements AttributeTypeServiceCompatibility {
+@SuppressWarnings("rawtypes")
+@OpenmrsProfile(openmrsPlatformVersion = "[2.2.0 - 2.3.*]")
+public class AttributeTypeServiceImpl2_2 implements AttributeTypeService {
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: This should be
```java
AttributeTypeServiceImpl2_2 extends AttributeTypeServiceImpl1_1 {
```

Comment (samuelmale) [reply api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: I don't agree with you on that. 

Why would we `extend` a handler? Basically we have no favor of inheritance we get in extending `AttributeTypeServiceImpl1_1`.

The question is why do we need this?

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java [hunk 14]
```diff
@@ -9,11 +9,11 @@
 import org.openmrs.attribute.BaseAttributeType;
 import org.openmrs.module.initializer.api.CsvLine;
 import org.openmrs.module.initializer.api.attributes.types.AttributeType;
-import org.openmrs.module.initializer.api.attributes.types.AttributeTypeLineProcessorCompatibilty;
+import org.openmrs.module.initializer.api.attributes.types.AttributeTypeCsvLineHandler;
 import org.openmrs.module.initializer.api.attributes.types.BaseAttributeTypeLineProcessor;
 
-@OpenmrsProfile(openmrsPlatformVersion = "2.2.0")
```
Comment (samuelmale) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java]: Btw, just out of curiosity; Why did you remove this?

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java [hunk 15]
```diff
@@ -15,15 +15,23 @@
 import org.openmrs.api.context.Context;
 import org.openmrs.attribute.BaseAttributeType;
 import org.openmrs.module.initializer.api.attributes.types.AttributeType;
-import org.openmrs.module.initializer.api.attributes.types.AttributeTypeServiceCompatibility;
+import org.openmrs.module.initializer.api.attributes.types.AttributeTypeService;
 
-@OpenmrsProfile(openmrsPlatformVersion = "2.2.0")
```
Comment (samuelmale) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: Samecase here?

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java (hunk 14 above)
Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeCsvLineHandlerImpl2_2.java]: @samuelmale this class was renamed, check on the renamed version of it.

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java (hunk 15 above)
Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: @samuelmale this class was also renamed, same remark.

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java [hunk 16]
```diff
@@ -1,133 +1,63 @@
...
+		return instance;
 	}
 	
 	@Override
-	public BaseAttributeType getAttributeTypeByUuid(String uuid, AttributeType attributeType) {
-		if (StringUtils.isEmpty(uuid) || attributeType == null) {
-			return null;
-		}
+	protected BaseAttributeType getByUuid(String uuid, AttributeTypeEnum attributeType) {
+
```
Comment (samuelmale) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: One question here. Did you just forget to reference the `super` implementation here? I guess this currently wont support `LOCATION`, `VISIT` and `PROVIDER` domains.

File: api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java [hunk 17]
```diff
@@ -1,133 +1,63 @@
...
-				if (existingProvAttTypes != null) {
-					for (ProviderAttributeType candidate : existingProvAttTypes) {
-						if (candidate.getName().equals(name)) {
-							return candidate;
-						}
-					}
-				}
-				return null;
-			
+	protected BaseAttributeType getByName(String name, AttributeTypeEnum typeEnum) {
```
Comment (samuelmale) [inline api-2.2/src/main/java/org/openmrs/module/initializer/attributes/types/AttributeTypeServiceImpl2_2.java]: Same case here @mks-d ?

**Comment by samuelmale** [comment]
>BaseAttributeTypeLineProcessorTest should not be an integration test, look at other domains for reference.

@mks-d What made the `BaseAttributeTypeLineProcessorTest` an integration test?
